sdk/core/typespec_client_core/src/http/mod.rs

```diff
@@ -104,6 +123,190 @@ impl UrlExt for Url {
+    ///
+    /// Both key and value accept types that can be converted to `Cow<'a, str>`, which includes
+    /// `&'a str`, `String`, and `Cow<'a, str>`. This avoids unnecessary allocations when using
+    /// string literals.
+    ///
+    /// Returns `&mut Self` to allow chaining multiple calls.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use typespec_client_core::http::{Url, UrlExt as _};
+    ///
+    /// let mut url: Url = "https://contoso.com?a=1".parse().unwrap();
+    /// url.query_builder()
+    ///     .append_pair("a", "2")
+    ///     .append_pair("b", "3")
+    ///     .build();
+    /// assert_eq!(url.as_str(), "https://contoso.com/?a=1&a=2&b=3");
+    /// ```
+    pub fn append_pair(
+        &mut self,
+        key: impl Into<std::borrow::Cow<'a, str>>,
+        value: impl Into<std::borrow::Cow<'a, str>>,
+    ) -> &mut Self {
+        let key = key.into();
+        let value = value.into();
+
+        if let Some((_, vals)) = self.values.iter_mut().find(|(k, _)| k == &key) {
+            vals.push(value);
+        } else {
+            self.values.push((key, vec![value]));
+        }
+
+        self
+    }
+
+    /// Sets a query parameter, overwriting any existing value with the same key.
+    ///
+    /// If the key already exists, all existing values for that key are replaced with the new value.
+    /// If the key doesn't exist, it's added to the end of the query parameters.
+    /// Use [`append_pair`](Self::append_pair) to add additional values without replacing existing ones.
+    ///
+    /// Both key and value accept types that can be converted to `Cow<'a, str>`, which includes
+    /// `&'a str`, `String`, and `Cow<'a, str>`. This avoids unnecessary allocations when using
+    /// string literals.
+    ///
+    /// Returns `&mut Self` to allow chaining multiple calls.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use typespec_client_core::http::{Url, UrlExt as _};
+    ///
+    /// let mut url: Url = "https://contoso.com?a=1&b=2".parse().unwrap();
+    /// url.query_builder()
+    ///     .set_pair("a", "new_value")
+    ///     .set_pair("c", "3")
+    ///     .build();
+    /// assert_eq!(url.as_str(), "https://contoso.com/?a=new_value&b=2&c=3");
+    /// ```
+    pub fn set_pair(
+        &mut self,
+        key: impl Into<std::borrow::Cow<'a, str>>,
+        value: impl Into<std::borrow::Cow<'a, str>>,
+    ) -> &mut Self {
+        let key = key.into();
+        let value = value.into();
+
+        if let Some((_, vals)) = self.values.iter_mut().find(|(k, _)| k == &key) {
+            // Replace existing value in-place (faster - no reordering)
+            *vals = vec![value];
+        } else {
+            // Add new key-value pair at the end
+            self.values.push((key, vec![value]));
+        }
+
+        self
+    }
+
+    /// Applies all the query parameter changes to the URL.
+    pub fn build(self) {
```

Comment:
I want to change this to `build(&mut self)` so that we can chain calls to `build()`. Let me know if you disagree.

Reply:
`build()` is generally last. What would you possibly chain after it? Builders are supposed to be init->setup->build (done).

Reply:
`append_pair` and `set_pair` return mutable references which prevents it though.